gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java

```diff
@@ -149,12 +118,124 @@ public abstract static class Builder {
 
     public abstract Builder setMtlsProvider(MtlsProvider mtlsProvider);
 
+    public abstract Builder setUsingGDCH(boolean usingGDCH);
+
+    public abstract Builder setResolvedEndpoint(String resolvedEndpoint);
+
+    public abstract Builder setResolvedUniverseDomain(String resolvedUniverseDomain);
+
+    abstract String serviceName();
+
+    abstract String universeDomain();
+
+    abstract String clientSettingsEndpoint();
+
+    abstract String transportChannelProviderEndpoint();
+
+    abstract String mtlsEndpoint();
+
+    abstract boolean switchToMtlsEndpointAllowed();
+
+    abstract MtlsProvider mtlsProvider();
+
+    abstract boolean usingGDCH();
+
+    abstract String resolvedUniverseDomain();
+
     abstract EndpointContext autoBuild();
 
+    private String determineUniverseDomain() {
+      // Do not set the universe domain for GDC-H
+      if (usingGDCH()) {
+        return null;
```

Comment:
Can we return GDU here instead of null? And make `resolvedUniverseDomain` not nullable? I see that `resolvedUniverseDomain` is used multiple places later, it is probably good now but may easily introduce null pointer exceptions later.

Reply:
The resolvedUniverseDomain is what is stored in the ClientContext and StubSettings (not the user configured universe domain). Following the GDC-H case above, it's possible to manually construct the StubSettings and ClientContext and each would construct the EndpointContext. If we set the resolvedUniverseDomain to the GDU by default, then it would fail every case as the Universe Domain can never be set for GDC-H (must be null).